posts/c-abi-changes-for-wasm32-unknown-unknown.md

```diff
@@ -0,0 +1,258 @@
+[leak-details]: https://github.com/rust-lang/rust/issues/115666
+
+## History of `wasm32-unknown-unknown`'s C ABI
+
+When the `wasm32-unknown-unknown` target [was originally added][inception] in
+2017 not much care was given to the exact definition of the `extern "C"` ABI at
+the time. In 2018 [an ABI definition wasm added just for wasm][orig-abi] and the
+target is still using this definition [to this day][current-abi]. This
+definitions has become more and more problematic over time and while some issues
+have been fixed, the root cause still remains.
+
+Notably this ABI definition does not match the [tool-conventions] definition of
+the C API, which is the current standard for how WebAssembly toolchains should
+talk to one another. Originally this non-standard definition was used for all
+WebAssembly based target except Emscripten, but [this changed in 2021][fix-wasi]
+where the WASI targets for Rust use a corrected ABI definition. Still, however,
+the non-standard definition remained in use for `wasm32-unknown-unknown`.
+
+The time has now come to correct this historical mistake and the Rust compiler
+will soon be using a correct ABI definition for the `wasm32-unknown-unknown`
+target. This means, however, that generated WebAssembly binaries will be
+different than before?
+
+## What is a WebAssembly C ABI?
+
+The definition of an ABI answers questions along the lines of:
+
+* What registers are arguments passed in?
+* What registers are results passed in?
+* How is a 128-bit integers passed as an argument?
+* How is a `union` passed as a return value?
+* When are parameters passed through memory instead of registers?
+* What is the size and alignment of a type in memory?
+
+For WebAssembly these answers are a little different than native platforms.
+WebAssembly does not have physical registers for example and functions must all
+be annotated with a type. What WebAssembly does have is types such as `i32`,
+`i64`, `f32`, and `f64`. This means that for WebAssembly an ABI needs to define
+how to represent values in these types.
+
+This is where the [tool-conventions] document comes in. That document provides a
+definition for how to represent primitives in C in the WebAssembly format, and
+additionally how function signatures in C are mapped to function signatures in
+WebAssembly. For example a Rust `u32` is represented by a WebAssembly `i32` and
+is passed directly as a parameter as a function argument. If the Rust structure
+`#[repr(C)] struct Pair(f32, f64)` is returned from a function then a return
+pointer is used which must have alignment 8 and size of 16 bytes.
+
+In essence the WebAssembly C ABI is acting as a bridge between C's type system
+and the WebAssembly type system. This includes details such as in-memory layouts
+and translations of a C function signature to a WebAssembly function signature.
+
+## How is `wasm32-unknown-unknown` non-standard?
```

Comment:
suggestion: if you'd like to link to a document describing this in its full extent (mostly), you can link to https://github.com/rust-diplomat/diplomat/blob/main/docs/wasm_abi_quirks.md

Reply:
I would very much like to link that, thank you!